File: riceCooker.ts
class RiceCooker {
    private isPowerOn: boolean;
    private contents: string[];
    private hasPower: boolean;
  
    constructor() {
      this.isPowerOn = false;
      this.contents = [];
      this.hasPower = false;
    }
  
    turnOn(): void {
      if (this.isPowerOn) {
        throw new Error("La rice cooker est déjà allumée.");
      }
  
      if (!this.hasPower) {
        throw new Error("La rice cooker ne peut pas être allumée car il n'y a pas de courant.");
      }
  
      this.isPowerOn = true;
      console.log("La rice cooker est allumée.");
    }
  
    turnOff(): void {
      if (!this.isPowerOn) {
        throw new Error("La rice cooker est déjà éteinte.");
      }
      this.isPowerOn = false;
      console.log("La rice cooker est éteinte.");
    }
  
    addContent(content: string): void {
      if (!this.isPowerOn) {
        throw new Error(
          "La rice cooker n'est pas allumée. Vous ne pouvez pas ajouter d'aliments."
        );
      }
      this.contents.push(content);
      console.log(`Vous avez ajouté "${content}" à la rice cooker.`);
    }
  
    cook(): void {
      if (!this.isPowerOn) {
        throw new Error(
          "La rice cooker n'est pas allumée. Vous ne pouvez pas faire cuire les aliments."
        );
      }
      console.log("La rice cooker est en train de cuire les aliments.");
      // Logique de cuisson des aliments
      console.log("La cuisson est terminée.");
    }
  
    setPower(hasPower: boolean): void {
      this.hasPower = hasPower;
    }
  }
  
  async function getUserInput(promptMessage: string): Promise<string | null> {
    const readline = require('readline');
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout
    });
  
    return new Promise((resolve) => {
      rl.question(promptMessage, (answer) => {
        rl.close();
        resolve(answer);
      });
    });
  }
  
  async function testRiceCooker(): Promise<void> {
    const riceCooker = new RiceCooker();
    let menu = 0;
  
    while (menu !== 5) {
      console.log("Veuillez choisir une action");
      console.log("1) Allumer la rice cooker");
      console.log("2) Ajouter un aliment");
      console.log("3) Cuire les aliments");
      console.log("4) Éteindre la rice cooker");
      console.log("5) Quitter");
  
      const userInput = await getUserInput("Choisissez une action : ");
      if (userInput !== null) {
        menu = parseInt(userInput, 10);
  
        if (menu === 1) {
          const hasPowerInput = await getUserInput("Y a-t-il du courant ? (Oui pour 'o', Non pour 'n') ");
          const hasPower = hasPowerInput === 'o';
          riceCooker.setPower(hasPower);
  
          try {
            riceCooker.turnOn();
          } catch (error) {
            console.log("Une erreur s'est produite : " + error.message);
          }
        } else if (menu === 2) {
          const content = await getUserInput("Veuillez ajouter un aliment : ");
          if (content !== null && content.length !== 0) {
            try {
              riceCooker.addContent(content);
            } catch (error) {
              console.log("Une erreur s'est produite : " + error.message);
            }
          }
        } else if (menu === 3) {
          try {
            riceCooker.cook();
          } catch (error) {
            console.log("Une erreur s'est produite : " + error.message);
          }
        } else if (menu === 4) {
          try {
            riceCooker.turnOff();
          } catch (error) {
            console.log("Une erreur s'est produite : " + error.message);
          }
        } else if (menu === 5) {
          console.log("Vous avez quitté le programme");
        } else {
          console.log("Veuillez choisir le bon menu");
        }
      } else {
        console.log("La valeur de l'entrée utilisateur est null.");
      }
    }
  }
  
  testRiceCooker();

File: src/RiceCooker.swift
import Foundation

class RiceCooker {
    var isPowerOn: Bool
    var contents: [String]
    var hasPower: Bool
    
    init() {
        isPowerOn = false
        contents = []
        hasPower = false
    }
    
    func turnOn() {
        if isPowerOn {
            fatalError("La rice cooker est déjà allumée.")
        }
        
        if !hasPower {
            fatalError("La rice cooker ne peut pas être allumée car il n'y a pas de courant.")
        }
        
        isPowerOn = true
        print("La rice cooker est allumée.")
    }
    
    func turnOff() {
        if !isPowerOn {
            fatalError("La rice cooker est déjà éteinte.")
        }
        
        isPowerOn = false
        print("La rice cooker est éteinte.")
    }
    
    func addContent(content: String) {
        if !isPowerOn {
            fatalError("La rice cooker n'est pas allumée. Vous ne pouvez pas ajouter d'aliments.")
        }
        
        contents.append(content)
        print("Vous avez ajouté \"\(content)\" à la rice cooker.")
    }
    
    func cook() {
        if !isPowerOn {
            fatalError("La rice cooker n'est pas allumée. Vous ne pouvez pas faire cuire les aliments.")
        }
        
        print("La rice cooker est en train de cuire les aliments.")
        // Logique de cuisson des aliments
        print("La cuisson est terminée.")
    }
    
    func setPower(hasPower: Bool) {
        self.hasPower = hasPower
    }
}

func getUserInput(promptMessage: String) -> String? {
    print(promptMessage, terminator: "")
    return readLine()
}

func testRiceCooker() {
    let riceCooker = RiceCooker()
    var menu = 0
    
    while menu != 5 {
        print("Veuillez choisir une action")
        print("1) Allumer la rice cooker")
        print("2) Ajouter un aliment")
        print("3) Cuire les aliments")
        print("4) Éteindre la rice cooker")
        print("5) Quitter")
        
        guard let userInput = getUserInput(promptMessage: "Choisissez une action : ") else {
            print("Une erreur s'est produite lors de la lecture de l'entrée utilisateur.")
            continue
        }
        
        menu = 0
        if let input = Int(userInput) {
            menu = input
        }
        
        switch menu {
        case 1:
            guard let hasPowerInput = getUserInput(promptMessage: "Y a-t-il du courant ? (Oui pour 'o', Non pour 'n') ") else {
                print("Une erreur s'est produite lors de la lecture de l'entrée utilisateur.")
                continue
            }
            
            let hasPower = hasPowerInput.lowercased() == "o"
            riceCooker.setPower(hasPower: hasPower)
            
            do {
                try riceCooker.turnOn()
            } catch {
                print("Une erreur s'est produite :", error)
            }
            
        case 2:
            guard let content = getUserInput(promptMessage: "Veuillez ajouter un aliment : ") else {
                print("Une erreur s'est produite lors de la lecture de l'entrée utilisateur.")
                continue
            }
            
            do {
                try riceCooker.addContent(content: content)
            } catch {
                print("Une erreur s'est produite :", error)
            }
            
        case 3:
            do {
                try riceCooker.cook()
            } catch {
                print("Une erreur s'est produite :", error)
            }
            
        case 4:
            do {
                try riceCooker.turnOff()
            } catch {
                print("Une erreur s'est produite :", error)
            }
            
        case 5:
            print("Vous avez quitté le programme")
            
        default:
            print("Veuillez choisir le bon menu")
        }
    }
}